### DISCUSSION TITLE
Discussion: mixed initializer styles

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#2616] [Language Ideas]
Currently both of these are valid initializers:
```cs
new StreamContent() { Headers = { ContentType = ct } };
new StreamContent() { Headers = { { "X-Header", h } } };
```
but not when they are mixed in a single initializer
```cs
new StreamContent() { Headers = { ContentType = ct, { "X-Header", h } } };
```
Looks like the parser already produce a valid tree for this, but the compiler eventually rejects it.

Could we relax the rules so that this is permitted?

**Comment by YairHalberstadt** [comment]
I find this a little confusing:

```csharp
using System.Collections.Generic;

public class C {
    public void M() {
        var list = new List<int>
        {
           Capacity = 100,
           1,
           2,
           3,
        };
    }
}
```

At first glance it looks like there's 4 items in the list, but actually theres 3.

However it could well be that if I got a little more used to this, I wouldn't find it so strange. But I think it would be nice to do some user studies on that.

**Comment by alrz** [comment]
> ```cs
>         var list = new List<int>
>         {
>            Capacity = 100,
>            1,
>            2,
>            3,
>         };
> ```

That case raises an ambiguity that currently is avoided precisely because you could either have a collection initializer *or* an object initializer, e.g. with the name `Capacity ` in scope, it will be assigned and the result will be added to the collection. My example is not ambiguous because we're passing more than one value.

**Comment by quinmars** [comment]
```csharp
int Capacity;
var list = new List<int>
{
    Capacity = 100,
    1,
    2,
    3,
};
```
Does not compile today. So it would not be a breaking change.

**Comment by YairHalberstadt** [comment]
But this does:

```csharp
int Capacity;
var list = new List<int>
{
    (Capacity = 100),
    1,
    2,
    3,
};
```

It might be a bad idea for them to have two separate meanings, but it's rare enough I don't think it matters.

**Comment by quinmars** [comment]
Those two separate meanings are already present, however, not at the same time.

```csharp
int Capacity;
var l1 = new List<int>
{
    Capacity = 100     // l1.Capacity = 100;
};
        
var l2 = new List<int>
{
    (Capacity = 100)   // l2.Add(100);
};
```

**Comment by mattwar** [comment]
This way its not ambiguous though.

```csharp
var list = new List<int>(new [] { 1, 2, 3 }) { Capacity = 100 };
```

Maybe we should work on turning it into this:

```csharp
var list = new List<int>({ 1, 2, 3 }) { Capacity = 100 };
```

Target-typed collection initializer anyone?

**Comment by vladd** [comment]
@mattwar What should be the type of `{ 1, 2, 3 }`? Is it a kind of array literal (as in `int[] a = { 1, 2, 3 }`), or it's contextually typed like lambda expressions?

**Comment by mattwar** [comment]
It was supposed to be an array initializer for a target-typed array/collection.  It was for the argument to List<T>(IEnumerable<T>).

**Comment by vladd** [comment]
@mattwar So a kind of array literal, right? Or you meant it to be interpreted differently in other contexts, e. g. `List<int> l = { 1, 2, 3 };` (so target-typed subclass of `IEnumerable<T>` defaulting to array)?

**Comment by mattwar** [comment]
For this case

```cs
List<int> x = { 1, 2, 3 };
```

would be interpreted as

```cs
List<int> x  = new List<int> { 1, 2, 3 };
```

But for cases where the target type is not a constructable class, but is assignable from array, like IEnumerable<T>, IReadOnlyList<T>, etc. 

```cs
IEnumerable<int> x = new int[] { 1, 2, 3 };
```

**Comment by vladd** [comment]
@mattwar Thank you for the explanation! Nice idea, is there an github issue to track its progress anywhere?

BTW I'd expect `IEnumerable<int> x = { 1, 2, 3 };` to produce actually an instance of some immutable collection rather then an array.

**Comment by mattwar** [comment]
@vladd not yet.

**Comment by alrz** [comment]
> For this case
> ```cs
> List<int> x = { 1, 2, 3 };
> ```
> would be interpreted as
> ```cs
> List<int> x  = new List<int> { 1, 2, 3 };
> ```

Using target-typed new, you can write
```cs
List<int> x  = new() { 1, 2, 3 };
```
If you don't mind `new()`, I think it covers the case where a target-typed initializer could be used.


Edit: Actually `new()` helps with some nasty syntax ambiguities if we want to keep `{}` on the literal. I think `[]` would be more appropriate, however with `new() {}` in place I'm not sure if that would worth it.
